syntax = "proto3";

message Offer {

    /*
    The max_amount values indicate maximum amount per individual transaction.
    The offering party is likely to reject larger transactions.

    After completing a transaction, the offering party may keep the offer open
    to accept more transactions under the same conditions, so the offer does not
    indicate the total amount available.

    The limit exchange rate is implied by the max_amount values.
    */

    message Asset {
        uint64 max_amount = 1;
        uint64 max_amount_divisor = 2; //e.g. 100 if max_amount is in EUR cent, or 100 000 000 if max_amount is in BTC satoshi
        string currency = 3;           //e.g. 'btc' for Bitcoin
        string exchange = 4;           //e.g. 'ln' for Lightning, or an exchange API URL
    }

    Asset bid = 1; //what is offered
    Asset ask = 2; //what must be paid

    //Transaction destination (e.g. Lightning node ID)
    string address = 3;

    //Transaction conditions
    message Condition {
        enum Key {
            UNKNOWN = 0;

            //For incoming LN funds:
            CLTV_EXPIRY_DELTA = 1; //incoming CLTV w.r.t. current block height

            //For custodial funds:
            LOCKED_TIMEOUT = 2;

            //Note: others might be added in the future.
        }
        Key key = 1;

        /*
        If a certain minimum or maximum is not needed, it can be set to the most
        extreme value, e.g. INT64_MIN or INT64_MAX.
        If an exact value is required, set both the minimum and maximum to that
        value.
        */
        sint64 min_value = 2;
        sint64 max_value = 3;
    }

    repeated Condition conditions = 4;
}

